src/OpenSpaceToolkit/Astrodynamics/NumericalSolver.cpp

```diff
@@ -373,6 +392,131 @@ Array<NumericalSolver::Solution> NumericalSolver::integrateDuration(
     return integrateTime(anInitialStateVector, 0.0, aDurationArray, aSystemOfEquations);
 }
 
+NumericalSolver::Solution NumericalSolver::integrateDuration(
+    const StateVector& anInitialStateVector,
+    const Real& aDurationInSeconds,
+    const SystemOfEquationsWrapper& aSystemOfEquations,
+    const Shared<EventCondition>& anEventCondition
+)
+{
+    if (aDurationInSeconds.isZero())
+    {
+        return {anInitialStateVector, 0.0};
+    }
+
+    NumericalSolver::StateVector aStateVector = anInitialStateVector;
+
+    // Ensure that the time step is the correct sign
+    const double signedTimeStep = getSignedTimeStep(aDurationInSeconds);
+
+    // TBI: Adapt this to any dense stepper type
+    auto stepper =
+        make_dense_output(absoluteTolerance_, relativeTolerance_, runge_kutta_dopri5<NumericalSolver::StateVector>());
+
+    // initialize stepper
+    double currentTime = 0.0;
+    stepper.initialize(aStateVector, currentTime, signedTimeStep);
+
+    // do first step
+    double previousTime;
+    std::tie(previousTime, currentTime) = stepper.do_step(aSystemOfEquations);
+    observeNumericalIntegration(stepper.current_state(), stepper.current_time());
+
+    bool conditionSatisfied = false;
+    Real currentValue = Real::Undefined();
+
+    const Size stateSize = anInitialStateVector.size();
+
+    Real previousValue = anEventCondition->evaluate(
+        Eigen::Map<const Eigen::VectorXd>(stepper.current_state().data(), stateSize), stepper.current_time()
+    );
+    NumericalSolver::StateVector currentState;
+
+    // account for integration direction
+    const auto checkTimeLimit = [&currentTime, &aDurationInSeconds]()
+    {
+        if (aDurationInSeconds > 0.0)
+        {
+            return currentTime < aDurationInSeconds;
+        }
+        return currentTime > aDurationInSeconds;
+    };
+
+    while (checkTimeLimit())
+    {
+        std::tie(previousTime, currentTime) = stepper.do_step(aSystemOfEquations);
+        currentState = stepper.current_state();
+
+        currentValue =
+            anEventCondition->evaluate(Eigen::Map<const Eigen::VectorXd>(currentState.data(), stateSize), currentTime);
+
+        conditionSatisfied = anEventCondition->isSatisfied(currentValue, previousValue);
+
+        if (conditionSatisfied)
+        {
+            break;
+        }
+
+        observeNumericalIntegration(currentState, currentTime);
+
+        previousValue = currentValue;
+    }
+
+    // TBI: Share information upstream on if the condition was satisfied
+    if (!conditionSatisfied)
+    {
+        stepper.calc_state(aDurationInSeconds, currentState);
+        return {currentState, aDurationInSeconds};
+    }
+
+    // Search for the exact time of the condition change
```

Comment:
So here we now the condition was met and we go further into determining the exact time with dichotomy because the integration scheme gives us an estimate based on stepper configuration, agreed?